tokio/benches/tcp.rs

```diff
@@ -182,76 +196,82 @@ mod transfer {
+        let addr = "127.0.0.1:0".parse::<SocketAddr>().unwrap();
 
         b.iter(move || {
-            let listener = TcpListener::bind(&addr).unwrap();
-            let addr = listener.local_addr().unwrap();
-
-            // Spawn a single future that accepts 1 connection, Drain it and drops
-            let server = listener
-                .incoming()
-                .into_future() // take the first connection
-                .map_err(|(e, _other_incomings)| e)
-                .map(|(connection, _other_incomings)| connection.unwrap())
-                .and_then(|sock| {
+            let result: io::Result<_> = block_on(async move {
+                let listener = TcpListener::bind(&addr).await?;
+                let addr = listener.local_addr().unwrap();
+
+                // Spawn a single future that accepts 1 connection, Drain it and drops
+                let server = async move {
+                    // take the first connection
+                    let sock = listener.incoming().into_future().await.0.unwrap()?;
                     sock.set_linger(Some(Duration::from_secs(0))).unwrap();
                     let drain = Drain {
                         sock,
                         chunk: read_size,
                     };
-                    drain
-                        .map(|_| ())
-                        .map_err(|e| panic!("server error: {:?}", e))
-                })
-                .map_err(|e| panic!("server err: {:?}", e));
+                    drain.map_ok(|_| ()).await
+                };
 
-            let client = TcpStream::connect(&addr)
-                .and_then(move |sock| Transfer {
+                let client = TcpStream::connect(&addr).and_then(move |sock| Transfer {
                     sock,
                     rem: MB,
                     chunk: write_size,
-                })
-                .map_err(|e| panic!("client err: {:?}", e));
+                });
 
-            server.join(client).wait().unwrap();
+                future::try_join(server, client).await
+            });
+            result.unwrap()
         });
     }
 
-    mod small_chunks {
+    pub mod small_chunks {
         use crate::prelude::*;
 
-        #[bench]
-        fn one_thread(b: &mut Bencher) {
+        pub fn one_thread(b: &mut Bencher<'_>) {
             super::one_thread(b, 32, 32);
         }
     }
 
-    mod big_chunks {
+    pub mod big_chunks {
         use crate::prelude::*;
 
-        #[bench]
-        fn one_thread(b: &mut Bencher) {
+        pub fn one_thread(b: &mut Bencher<'_>) {
             super::one_thread(b, 1_024, 1_024);
         }
     }
 }
+
+use prelude::*;
+
+fn bench_tcp(c: &mut Criterion) {
+    c.bench_function("connect_churn/one_thread", connect_churn::one_thread);
+    c.bench_function("connect_churn/two_threads", connect_churn::two_threads);
+    c.bench_function("connect_churn/multi_threads", connect_churn::multi_threads);
```

Comment:
I think this could maybe be converted to a Criterion parameterized benchmark parameterized over the number of threads, but it's fine to do that in a follow-up.